phdi/fhir/tabulation/tables.py

```diff
@@ -50,36 +50,190 @@ def _apply_selection_criteria(
+
+    First, resources that are associated with one another in the
+    provided schema are grouped together. For example, a Patient
+    resource has a field `generalPractitioner` which holds a FHIR
+    `reference` dictionary giving the resource ID of the patient's
+    traditional doctor. For each table, one type of resource serves
+    as the "anchor" resource, which will define the number of rows
+    in the table. Other resources may be referrenced by either
+    "forwards" (an anchor resource has a reference to the other
+    resource, i.e. the Practitioner in the example above) or
+    "backwards" association (another resource has a reference to the
+    anchor resource, i.e. an Observation resource for a medical Exam
+    has a `subject` field holding a reference to an anchor Patient).
+    These related resources are stored in a data structure
+    accessible using the anchor resource's FHIR ID for efficient lookup.
+
+    Second, the aggregated resources are parsed for value extraction.
+    A second pass is made over just the collected anchor resources, and
+    using the reference mappings made in the first pass, all other
+    requested data from other resources is looked up and extracted.
+    The resulting collection of values is stored in a list in accordance
+    with the provided schema's column headers, and these lists are
+    appended as rows to the final tabulated data set.
+
+    The resulting list of lists serves as the output data table. The
+    first entry is a list of headers for the new table, and
+    subsequent lists are each an extracted data element represented
+    in a tabular format. The returned value of this function is a
+    dictionary whose keys are the names of tables (as given in the
+    provided schema) and whose values are the output tables.
+
+    :param extracted_data: The bundle of FHIR data to tabulate. Must
+      include all resources of each type to reference.
+    :param schema: A schema of columns and values to apply to the
+      extracted data, separated by resource type.
+    :return: A dictionary mapping table names to lists of lists.
+      The first list in the return value is a list of headers
+      serving as the columns, and all subsequent lists are rows in
+      the table.
+    """
+
+    # First pass: build mapping of references for easy lookup
+    ref_directions = _get_reference_directions(schema)
+    ref_dicts = _build_reference_dicts(extracted_data, ref_directions)
+
+    data = {}
+    for table_name in ref_dicts.keys():
+        # Get the columns from the schema so we always iterate through
+        # them in a consistent order
+        column_items = (
+            schema.get("tables", {}).get(table_name, {}).get("columns", {}).items()
+        )
+        headers = [c[0] for c in column_items]
+        data[table_name] = [headers]
+        anchor_type = (
+            schema.get("tables", {}).get(table_name, {}).get("resource_type", "")
+        )
```

Comment:
Just a minor refactor suggestion for readability.
```suggestion
    for table_name, table_params in schema["tables"].items():
        # Get the columns from the schema so we always iterate through
        # them in a consistent order
        column_items = table_params.get("columns", {}).items()
        headers = [column_name for column_name, _ in column_items]
        data[table_name] = [headers]
        anchor_type = table_params.get("resource_type", "")
```

Reply:
I like this! Will implement.